d3.js:1378 Error: <circle> attribute cx: Expected length, "NaN".

I worked around the error I reported earlier related to npm install by cloning the repo and linking it into my project. That is working fine.

I have a trivially simple example that looks like this:

```
import * as d3 from 'd3';
import eventDrops from 'event-drops/dist/eventDrops';

    const chart = eventDrops()
      .start(new Date(2017, 1, 1))
      .end(new Date(2017, 4, 31))
      .eventColor("red")
      .margin({ top: 60, left: 0, bottom: 40, right: 50 })
      .eventLineColor("blue")
      .labelsWidth(100)
    ;

    d3.select("#eventDrops")
      .datum([
        {name: "Bob", data: [new Date(), new Date(2017, 3, 18)]},
        {name: "Sally", data: [new Date(2017, 4, 3), new Date(2017, 5, 22)]},
      ])
      .call(chart);
```

The chart displays, but whenever I drag the timeline, I get

> d3.js:1378 Error: <circle> attribute cx: Expected length, "NaN".

And the drops disappear.

I will continue to try to isolate this problem, and/or create a simple reproducible test case (at the moment, this is inside my large Angular app), but is there any obvious reason this error would be appearing?

--- a/src/zoom.js
+++ b/src/zoom.js
@@ -5,18 +5,20 @@ import { boolOrReturnValue } from './drawer/xAxis';
 export const onRequestAnimationFrameFactory = (
     container,
     configuration,
-    callback
+    callback,
+    sumDataCount,
+    scalingFunction,
+    data
 ) =>
-    (sumDataCount, scalingFunction, data) =>
-        () => {
-            container
-                .selectAll('.drop-line')
-                .selectAll('.drop')
-                .attr('cx', d => scalingFunction(configuration.date(d)));
+    () => {
+        container
+            .selectAll('.drop-line')
+            .selectAll('.drop')
+            .attr('cx', d => scalingFunction(configuration.date(d)));
 
-            sumDataCount(data);
-            callback(data);
-        };
+        sumDataCount(data);
+        callback(data);
+    };
 
 export default (
     container,
@@ -50,7 +52,10 @@ export default (
         }
 
         global.requestAnimationFrame(
-            onRequestAnimationFrameFactory(container, configuration, callback)(
+            onRequestAnimationFrameFactory(
+                container,
+                configuration,
+                callback,
                 sumDataCount,
                 scalingFunction,
                 data

--- a/src/zoom.js
+++ b/src/zoom.js
@@ -2,9 +2,40 @@ import debounce from 'debounce';
 import labels from './drawer/labels';
 import { boolOrReturnValue } from './drawer/xAxis';
 
-export default (container, dimensions, scales, configuration, callback) => {
-    const onZoom = (data) => {
+export const onRequestAnimationFrameFactory = (
+    container,
+    configuration,
+    callback
+) =>
+    (sumDataCount, scalingFunction, data) =>
+        () => {
+            container
+                .selectAll('.drop-line')
+                .selectAll('.drop')
+                .attr('cx', d => scalingFunction(configuration.date(d)));
+
+            sumDataCount(data);
+            callback(data);
+        };
+
+export default (
+    container,
+    dimensions,
+    scales,
+    configuration,
+    callback = () => {}
+) => {
+    const onZoom = data => {
         const scalingFunction = d3.event.transform.rescaleX(scales.x);
+
+        const sumDataCount = debounce(
+            labels(
+                container.select('.labels'),
+                { x: scalingFunction },
+                configuration
+            ),
+            100
+        );
 
         if (boolOrReturnValue(configuration.hasTopAxis, data)) {
             container
@@ -18,27 +49,13 @@ export default (container, dimensions, scales, configuration, callback) => {
                 .call(d3.axisBottom().scale(scalingFunction));
         }
 
-        const sumDataCount = debounce(
-            labels(
-                container.select('.labels'),
-                { x: scalingFunction },
-                configuration
-            ),
-            100
+        global.requestAnimationFrame(
+            onRequestAnimationFrameFactory(container, configuration, callback)(
+                sumDataCount,
+                scalingFunction,
+                data
+            )
         );
-
-        global.requestAnimationFrame(() => {
-            container
-                .selectAll('.drop-line')
-                .selectAll('.drop')
-                .attr('cx', d => scalingFunction(new Date(d.date)));
-
-            sumDataCount(data);
-
-            if (callback) {
-                callback(data);
-            }
-        });
     };
 
     const zoom = d3
@@ -48,5 +65,6 @@ export default (container, dimensions, scales, configuration, callback) => {
         .on('end', configuration.zoomend);
 
     container.call(zoom);
+
     return zoom;
 };

--- a/test/karma/eventDrops.js
+++ b/test/karma/eventDrops.js
@@ -66,9 +66,10 @@ describe('eventDrops', () => {
         const zoom = require('../../src/zoom');
         const data = [ { name: 'foo', data: [new Date()] }];
 
-        const test = (zoomable, expectedZoomableBehavior) => {
-            zoom.default = jasmine.createSpy();
+        spyOn(zoom, 'default').and.callThrough();
 
+        const test = (zoomable, expectedZoomableBehavior) => {
+            zoom.default.calls.reset();
             const div = document.createElement('div');
 
             const chart = eventDrops().zoomable(zoomable);

--- a/test/karma/zoom.js
+++ b/test/karma/zoom.js
@@ -19,8 +19,14 @@ describe('Zoom', () => {
         const drops = container.selectAll('.drop').data(data);
 
         const scalingFunction = x => x;
-        const onRequestAnimationFrame = onRequestAnimationFrameFactory(container, configuration, () => {});
-        onRequestAnimationFrame(() => {}, scalingFunction, data)();
+        onRequestAnimationFrameFactory(
+            container,
+            configuration,
+            () => {},
+            () => {},
+            scalingFunction,
+            data
+        )();
 
         // .attr getter returns result of function with first non null data specified
         // @see http://devdocs.io/d3~4/d3-selection#selection_attr

--- a/test/karma/zoom.js
+++ b/test/karma/zoom.js
@@ -1,0 +1,29 @@
+import { onRequestAnimationFrameFactory } from '../../src/zoom';
+
+describe('Zoom', () => {
+    it('should rescale using custom date using configuration.date function', () => {
+        const configuration = {
+            date: d => d.customDate,
+        };
+
+        const div = document.createElement('div');
+        div.innerHTML = `
+            <div class="drop-line">
+                <div class="drop"></div>
+            </div>
+        `;
+
+        const data = [{ id: 1, customDate: '2017-01-12' }];
+        const container = d3.select(div);
+
+        const drops = container.selectAll('.drop').data(data);
+
+        const scalingFunction = x => x;
+        const onRequestAnimationFrame = onRequestAnimationFrameFactory(container, configuration, () => {});
+        onRequestAnimationFrame(() => {}, scalingFunction, data)();
+
+        // .attr getter returns result of function with first non null data specified
+        // @see http://devdocs.io/d3~4/d3-selection#selection_attr
+        expect(drops.attr('cx')).toBe('2017-01-12');
+    });
+});

